nall/directory.hpp

```diff
@@ -267,13 +267,23 @@ inline auto directory::copy(const string& source, const string& target) -> bool
   }
 #else
   inline auto directoryIsFolder(DIR* dp, struct dirent* ep) -> bool {
+#if defined(PLATFORM_HAIKU)
+    struct stat sp = {0};
+    stat(ep->d_name, &sp);
+    if S_ISDIR(sp.st_mode) return true;
+    if (S_ISLNK(sp.st_mode) || not S_ISREG(sp.st_mode)) {
+      fstatat(dirfd(dp), ep->d_name, &sp, 0);
+      return S_ISDIR(sp.st_mode);
+    }
+#else // not PLATFORM_HAIKU
```

Comment:
It seems that POSIX only mandates the `d_name` and `d_ino` fields, not `d_type`. Rather than this being a special-case for `PLATFORM_HAIKU` it should probably be the default implementation for POSIXy OSs, and the existing implementation should be behind `#if defined(PLATFORM_MACOS) || defined(PLATFORM_LINUX) || defined(PLATFORM_BSD)` since those are the platforms we happen to know have this extra feature.

We should probably also have a fallback to the above code if `d_type == DT_UNKNOWN` (for when the filesystem doesn't support `d_type`) too, but you don't have to do that in this PR, I'm just thinking out loud.

Reply:
Indeed, d_type is the most common thing I need to look for when porting things to Haiku. I will turn the ifdef around as you proposed, it will also help for other future ports (another one of my pet peeves is porting libretro cores to Solaris, and this kind of things happens there as well).